File: linked_lists/questions/question_2.1/problem_2.1.rb
require_relative './../../../custom_data_structures/ruby_data_structures/linked_list.rb'

# class Linked_list

#   def remove_duplicates
#     values_present = {}
#     runner = head
#     while runner
#       if values_present[runner.value]
#         temp = runner.next
#         remove_node runner
#         runner = temp
#       else
#         values_present[runner.value] = true
#       end
#       runner = runner.next
#     end
#   end

#   def remove_duplicates_no_buffer
#     current = head
#     while current
#       find_and_delete_after current
#       current = current.next
#     end
#   end

#   private
#   def find_and_delete_after current
#     runner = current.next
#     while runner
#       (remove_node runner) if runner.value == current.value
#       runner = runner.next
#     end
#   end
# end

class LinkedList
  def remove_duplicates
    return unless @head && @head.next
    parent = @head
    runner = @head.next
    presence_hash = {@head.value => true}
    while runner
      if presence_hash[runner.value]
        parent.next = runner.next
      else
        presence_hash[runner.value] = true
        parent = parent.next
      end
      runner = runner.next
    end
  end
end

File: linked_lists/questions/question_2.1/problem_2.1.js
// I'm not running this in a browser.  just running it in node from my terminal.  I can't figure out how to 'include' the constructor and prototype functions in this file.
// i'll copy and paste this code into the linked_list.js file until I learn the proper include syntax
LinkedList.prototype.removeDuplicates = function(){
  var duplicatesArray = []
  var runner = this.head
  while (runner){
    // debugger
    if (duplicatesArray.includes(runner.value)){
      this.deleteNode(runner)
      if (!runner.next){
        break;
      }
    } else {
      duplicatesArray.push(runner.value)
      runner = runner.next
    }
  }
}

LinkedList.prototype.removeDuplicatesNoBuffer = function(){
  var current = this.head
  while (current){
    findAndRemoveDups (current, this)
    current = current.next
  }
}

var findAndRemoveDups = function(targetNode, classInstance){
  var runner = targetNode.next
  while (runner){
    if (runner.value == targetNode.value){ classInstance.removeNode(runner)}
    runner = runner.next
  }
}

var list = new LinkedList()
for (var val of [1,2,3,2,1,3,5,6,7,5,4,2,9]){
  list.push(val)
}
list.viewList()
console.log('---------------------------------------------')
list.removeDuplicates()
console.log('---------------------------------------------')
list.viewList()
